src/vsphere_cpi/lib/cloud/vsphere/attach_tag_to_vm.rb

```diff
@@ -0,0 +1,172 @@
+            next
+          end
+          tag_hash[vm_config_tag["category"]] ||= []
+          tag_hash[vm_config_tag["category"]] << vm_config_tag["tag"]
+        end
+        tag_hash
+      end
+
+      def attach_single_tag(vm_mob_id, tag_id )
+        tag_assoc_info = VSphereAutomation::CIS::CisTaggingTagAssociationAttach.new
+        tag_assoc_info.object_id = VSphereAutomation::CIS::VapiStdDynamicID.new
+        tag_assoc_info.object_id.id = vm_mob_id
+        tag_assoc_info.object_id.type = 'VirtualMachine'
+        tag_assoc_api.attach(tag_id,tag_assoc_info)
+      end
+
+      def attach_multi_tags(vm_mob_id, tag_ids, category_name, target_category_id)
+        single_cardinality = single_category_cardinality?(target_category_id)
+        if single_cardinality
+          logger.warn( "Tag category '#{category_name}' does not support 'Multiple Cardinality', skip tagging this category to vm." )
+          return
+        end
+        multi_tag_assoc_info = VSphereAutomation::CIS::CisTaggingTagAssociationAttachMultipleTagsToObject.new
+        multi_tag_assoc_info.object_id = VSphereAutomation::CIS::VapiStdDynamicID.new
+        multi_tag_assoc_info.object_id.id = vm_mob_id
+        multi_tag_assoc_info.object_id.type = 'VirtualMachine'
+        multi_tag_assoc_info.tag_ids = tag_ids
+        tag_assoc_api.attach_multiple_tags_to_object(multi_tag_assoc_info)
+      end
+
+      def attach_tags(vm_mob_id, vm_config_tags, vm_config_name)
+        tag_hash = create_tag_hash(vm_config_tags)
+        if tag_hash.empty?
+          logger.warn( "No valid category-tag pair in cloud config, finish attaching tags." )
+          return
+        end
+        category_ids = tagging_category_api.list.value
+        if category_ids.empty?
+          logger.warn( "No category exist on vCenter, skip attaching tags." )
+          return
+        end
+
+        tag_hash.keys.each do |category_name|
+          begin
+            target_category_id = retrieve_category_id(category_name, category_ids)
+            if target_category_id.nil?
+              logger.warn( "Tag category '#{category_name}' does not exit in vCenter, skip attaching this category to vm." )
+              next
+            end
+            vm_association = vm_association?(target_category_id)
+            unless vm_association
+              logger.warn( "Tag category '#{category_name}' is not associated with object type: 'Virtual Machine', skip attaching this category to vm." )
+              next
+            end
+            tag_array = tag_hash[category_name]
+            logger.warn( "Duplicated tags found in tag category '#{category_name}', deduplicated and continue" ) if tag_array.uniq!
+
+            tag_id_list = tagging_tag_api.list_tags_for_category(target_category_id).value
+            tag_ids = tag_array.inject([]) do |tag_ids, tag_name|
+              tag_ids << retrieve_tag_id(tag_name, tag_id_list)
+            end.compact
+
+            if tag_ids.empty?
```

Comment:
This looks more like a `case ... when ... end` situation than if statements.

Reply:
Updated the code with `case ... when ... end` .